lib/src/clojure_lsp/feature/diagnostics.clj

```diff
@@ -128,48 +131,49 @@
           (get-in db [:clj-depend-violations (symbol namespace)]))))
 
 (defn find-diagnostics [^String uri db]
-  (let [filename (shared/uri->filename uri)
-        kondo-level (settings/get db [:linters :clj-kondo :level])
+  (let [kondo-level (settings/get db [:linters :clj-kondo :level])
         depend-level (settings/get db [:linters :clj-depend :level] :info)]
-    (if (shared/jar-file? filename)
+    (if (shared/jar-file? uri)
       []
       (cond-> []
         (not= :off kondo-level)
-        (concat (kondo-findings->diagnostics filename :clj-kondo db))
+        (concat (kondo-findings->diagnostics uri :clj-kondo db))
 
         (not= :off depend-level)
-        (concat (clj-depend-violations->diagnostics filename depend-level db))))))
+        (concat (clj-depend-violations->diagnostics uri depend-level db))))))
 
 (defn ^:private publish-diagnostic!* [{:keys [diagnostics-chan]} diagnostic]
   (async/put! diagnostics-chan diagnostic))
 
 (defn ^:private publish-all-diagnostics!* [{:keys [diagnostics-chan]} diagnostics]
   (async/onto-chan! diagnostics-chan diagnostics false))
 
+(defn ^:private diagnostics-of-uri [uri db]
+  {:uri uri
+   :diagnostics (find-diagnostics uri db)})
+
+(defn ^:private empty-diagnostics-of-uri [uri]
+  {:uri uri
+   :diagnostics []})
+
 (defn publish-diagnostics! [uri {:keys [db*], :as components}]
-  (publish-diagnostic!* components {:uri uri
-                                    :diagnostics (find-diagnostics uri @db*)}))
+  (publish-diagnostic!* components (diagnostics-of-uri uri @db*)))
 
-(defn publish-all-diagnostics! [paths {:keys [db*], :as components}]
+(defn publish-all-diagnostics! [uris {:keys [db*], :as components}]
   (let [db @db*]
     (publish-all-diagnostics!*
       components
-      (eduction (map io/file)
-                (mapcat file-seq)
-                (map #(.getAbsolutePath ^java.io.File %))
-                (map #(shared/filename->uri % db))
-                (remove #(= :unknown (shared/uri->file-type %)))
```

Comment:
I'd like a more careful review of this. The `crawler` used to pass `publish-all-diagnostics!` the `:source-paths` directly, which is just a list of directories like `["src", "test"]`. We used `file-seq` to figure out what files are in those directories, removed the non-Clojure files, and published their diagnostics. But this is overkill I think. By the time `publish-all-diagnostics!` is called, we've already done the initial project analysis. During that process clj-kondo has figured out which files are in those directories, and analyzed whatever it can. Those files are all represented in the `:findings` map in the db. So we might as well just publish the diagnostics we have for those files—i.e., all the findings we have. I've switched the code in `crawler` to do that and confirmed that the other uses of `publish-all-diagnostics!` still work as expected. Here I've removed the `file-seq` related code and assumed we'll be passed URIs of files, not paths. I've left in the `(remove #(= :unknown (shared/uri->file-type %)))`, but I think that could be removed too. How would we have findings for an `:unknown` file? So, 1) does it seem safe to remove `file-seq`? And 2) do we need the `(remove :unknown ...)` code?

Reply:
Ah, I didn't realize didChangeWatchedFiles could include directories, but the spec does mention that. Hmm... I'll look into restoring some of this.

Reply:
I added `file-seq` back in, but only for `didChangeWatchedFiles`. I did that a higher level, at the beginning of `clojure-lsp.feature.file-management/did-change-watched-files`, so that `analyze-watched-files!` and `files-deleted` both receive file URIs, not directory URIs. It looks like the code in those functions (including, but not limited to `publish-all-diagnostics!`) expects file URIs.